randovania/game_description/game_database_view.py

```diff
@@ -0,0 +1,304 @@
+        """
+
+    @abc.abstractmethod
+    def node_from_pickup_index(self, index: PickupIndex) -> PickupNode:
+        """
+        Returns the PickupNode with the given index.
+        :raises: KeyError if it doesn't exist
+        """
+
+    @abc.abstractmethod
+    def area_from_node(self, node: Node) -> Area:
+        """
+        Returns the Area that contains the given node.
+        :raises: KeyError if it doesn't exist
+        """
+
+
+class GameDatabaseViewProxy(GameDatabaseView):
+    """
+    A GameDatabaseView, implemented by delegating all calls to another GameDatabaseView.
+    Intended to be used to overwrite specific functions.
+    """
+
+    def __init__(self, original: GameDatabaseView):
+        self._original = original
+
+    @override
+    def node_iterator(self) -> Iterable[tuple[Region, Area, Node]]:
+        return self._original.node_iterator()
+
+    @override
+    def iterate_nodes_of_type[NodeT: Node](self, node_type: type[NodeT]) -> Iterator[tuple[Region, Area, NodeT]]:
+        return self._original.iterate_nodes_of_type(node_type)
+
+    @override
+    def node_by_identifier(self, identifier: NodeIdentifier) -> Node:
+        return self._original.node_by_identifier(identifier)
+
+    @override
+    def assert_pickup_index_exists(self, index: PickupIndex) -> None:
+        return self._original.assert_pickup_index_exists(index)
+
+    @override
+    def create_resource_collection(self) -> ResourceCollection:
+        return self._original.create_resource_collection()
+
+    @override
+    def default_starting_location(self) -> NodeIdentifier:
+        return self._original.default_starting_location()
+
+    @override
+    def get_dock_types(self) -> list[DockType]:
+        return self._original.get_dock_types()
+
+    @override
+    def get_dock_weakness(self, dock_type_name: str, weakness_name: str) -> DockWeakness:
+        return self._original.get_dock_weakness(dock_type_name, weakness_name)
+
+    @override
+    def get_resource_database_view(self) -> ResourceDatabaseView:
+        return self._original.get_resource_database_view()
+
+    @override
+    def get_pickup_database(self) -> PickupDatabase:
+        return self._original.get_pickup_database()
+
+    @override
+    def get_victory_condition(self) -> Requirement:
+        return self._original.get_victory_condition()
+
+    @override
+    def pickup_nodes_with_feature(self, feature: HintFeature) -> tuple[PickupNode, ...]:
+        return self._original.pickup_nodes_with_feature(feature)
+
+    @override
+    def node_from_pickup_index(self, index: PickupIndex) -> PickupNode:
+        return self._original.node_from_pickup_index(index)
+
+    @override
+    def area_from_node(self, node: Node) -> Area:
+        return self._original.area_from_node(node)
+
+
+def typed_node_by_identifier[NodeT: Node](game: GameDatabaseView, i: NodeIdentifier, t: type[NodeT]) -> NodeT:
```

Comment:
any reason this can't be a bound method on `GameDatabaseView`? possibly decorated as `@final`?

Reply:
maybe the solution is just to have it all be `@final`. I was thinking we have the GameDatabaseView as the minimum API, and then something else to build higher level APIs

Right now I'm leaving `node_by_identifier` being re-implemented in GameDescription, but the goal is that'd be just iterate and search.